### ISSUE TITLE
[ info ] there is a custom node to cast, mix and save GGUF models !

### ISSUE THREAD

**Issue by JorgeR81** [city96/ComfyUI-GGUF#78]
https://github.com/chrisgoringe/cg-mixed-casting

**Comment by city96** [comment]
Pretty neat idea, looks like they've keeping the first/last layer in a higher precision. llama.cpp has [similar logic](https://github.com/ggerganov/llama.cpp/blob/66b039a5011522b6a61495eea2a9862601e169f7/src/llama.cpp#L16360-L16362) as well for some of the _M quants (we currently don't use this though since we'd have to enumerate the layer names to find the actual block count, which would need extra logic with conv models that might have up/down/mid blocks + weird diffusers naming for the keys).

Would be interesting to see how much this can help just plain fp8 as well, with maybe some logic for keeping the biases/some other important weights in FP16 for a few hundred MB size increase (maybe it already has that I just missed it).

@chrisgoringe

**Comment by chrisgoringe** [comment]
Hey, thanks for the mention. Like I say, the GGUF bits are heavily based on your code :). 

It can be used to do a mixed fp8/fp16 model, but I've found fp8 to be significantly less accurate than Q8_0 (although faster). More below on accuracy!

In general I keep the in and out layers in bf16 (mostly because they are small, and complicated!). I also generally keep the first couple of double block layers, and the last double block layer in bf16. The last double block seems particularly significant, perhaps because of the merge of the txt and img streams after it? The single block layers are *much* less sensitive to changes.

I haven't considered keeping the biases at different precision to the weights - that's a really interesting idea. You think they are more significant?

What I'd really like is for more of the quantisations to be supported in quantise, not just dequantise. Things like Q4_K, since Q4_1 is pretty outdated as I understand it. Don't suppose you're working on that?

I'm currently working on objectively quantifying the impact on the accuracy of any specific quantization. What I've done is run a few hundred random prompts (from the civit.ai API of popular images generated using FLUX), and stash the results on hugging face (they are BIG!).

Then you can just make any change to the model, and run the inputs through, and do an MSE calculation to see how much that change has changed the output.

Love to talk more if you are interested!

**Comment by city96** [comment]
>It can be used to do a mixed fp8/fp16 model, but I've found fp8 to be significantly less accurate than Q8_0 (although faster). More below on accuracy!

I meant specifically fp8+fp16/bf16 mixed as to have something that could (potentially) work in training backends/etc without having to add support, assuming they don't just do a `.to` on the model weights. Q8_0 is definitely more accurate.

>The last double block seems particularly significant, perhaps because of the merge of the txt and img streams after it? The single block layers are much less sensitive to changes.

This is actually pretty good info, we can add this logic to our M quant logic (possibly).

>I haven't considered keeping the biases at different precision to the weights - that's a really interesting idea. You think they are more significant?

I think llama.cpp does it partially because they're very small, and not dequantizing them probably provides at least some speed benefit and adds less complexity (though the per-weight throughput for small tensors was high enough to not matter).

They're using FP32 for most small tensors as well. According to comfy not quantizing them makes sense, though the stable-diffusion.cpp team said they didn't see much change from doing this. It's give or take 50MBs so your call I guess.

>What I'd really like is for more of the quantizations to be supported in quantize, not just dequantize. Things like Q4_K, since Q4_1 is pretty outdated as I understand it. Don't suppose you're working on that?

I assume you're using the gguf-py quantize function so you'd have to ask the llama.cpp people to implement them. I've moved all the quantization code to C++ in this repository for simplicity and maintainability and the pytorch ones are just ports of the numpy ones in the library. You can technically hook into the ggml.dll with ctypes but this is absolute ass (or maybe I'm stupid) and I haven't gotten it to work reliably even for dequantization lmao.

>Then you can just make any change to the model, and run the inputs through, and do an MSE calculation to see how much that change has changed the output.

Would MSE work here? Wouldn't you want something like feature loss (i.e. generate a pooled clip embedding for the image and compare that)? Just MSE would change a lot if an object is a few inches off on a table somewhere, while changing less if some gray shadow on a white wall suddenly disappears.

**Comment by chrisgoringe** [comment]
> > It can be used to do a mixed fp8/fp16 model, but I've found fp8 to be significantly less accurate than Q8_0 (although faster). More below on accuracy!
> 
> I meant specifically fp8+fp16/bf16 mixed as to have something that could (potentially) work in training backends/etc without having to add support, assuming they don't just do a `.to` on the model weights. Q8_0 is definitely more accurate.

Right, that makes sense. Might want to look at the approach taken by bitsandbytes. I did take a quick look at trying to do a quantisation that allowed the gradient to be kept in f32 while the forward step (and update) was done with a quant, but my brain melted.

> I assume you're using the gguf-py quantize function so you'd have to ask the llama.cpp people to implement them. ...

OK, thanks

> > Then you can just make any change to the model, and run the inputs through, and do an MSE calculation to see how much that change has changed the output.
> 
> Would MSE work here? Wouldn't you want something like feature loss (i.e. generate a pooled clip embedding for the image and compare that)? Just MSE would change a lot if an object is a few inches off on a table somewhere, while changing less if some gray shadow on a white wall suddenly disappears.

I think they are answering different questions. I'm measuring the error in the final hidden state before it is projected out to the latent space as a noise prediction - quite a long way from an actual image :). This allows a direct comparison of the impact of different ways of modifying the model (eg quantising)

The question of how errors in the final hidden state result in meaningful changes to the image is more involved (and arguably more subjective). 

I'm assuming that when considered over a large number of different prompts etc., there is a monotonic relationship between errors in the final hidden state and reduced 'quality' of the final result.

**Comment by city96** [comment]
>I think they are answering different questions. I'm measuring the error in the final hidden state before it is projected out to the latent space as a noise prediction - quite a long way from an actual image :). This allows a direct comparison of the impact of different ways of modifying the model (eg quantising)

Ohh, smart. Yeah that would be super interesting to see considering we'd get a quantifiable metric as you say. I do wonder if random changes early on would have a larger/cascading effect on the hidden states later on. I guess you could check at multiple points but that'd be a bit too painful to do. We're testing an image model not drilling for oil lol.

**Comment by chrisgoringe** [comment]
😂 I'm a total data junkie. So I'm currently running an automated test that is checking the effect on the overall stack of different quants in different blocks.

I do also have a (huge!) dataset with the hidden states after each layer. In a separate project I'm trying reducing the size of the feed forward layers by removing the internal states that are least frequently used, and then retaining the rest of the matrix to reduce the error...

**Comment by chrisgoringe** [comment]
So... here's a page with a stupid amount of data in it... 

https://github.com/chrisgoringe/flux-poke/blob/main/docs/casting_cost.md

Initial takeaways:
- bitsandbytes performs poorly
- the later layers are much less sensitive than the earlier ones (though not monotonically, and even some variation between quants)
- quantise the biases, don't quantise the biases, whatever, no real difference
- the optimum accuracy v. memory is going to be a mixture

Still to do:
- complete the table, adding bfloat8 as well
- do some speed measurements
- turn the table into a mechanism to choose the optimum quantisation for any given size
